File: packages/sdk/src/internal/signaling/protocol.ts
/**
 * Raven's signaling wire protocol, as the client sees it.
 *
 * Mirrors `apps/api/src/modules/signaling/signaling.constants.ts` and
 * `interfaces/signaling-message.interface.ts`. Written out by hand rather
 * than imported, so the SDK carries no dependency on the server package.
 * Three client implementations speak this (web, React Native, Flutter) and
 * the contract has to stand on its own.
 *
 * Warning if you're reading older code: most of the message *names*
 * survived the move off the old full-mesh protocol, but their meanings
 * didn't. `targetParticipantId` is gone. A client has exactly one peer
 * now: the SFU serving its room.
 */

export const ClientMessageType = {
  ROOM_JOIN: 'room.join',
  ROOM_LEAVE: 'room.leave',
  SDP_ANSWER: 'sdp.answer',
  SDP_OFFER: 'sdp.offer',
  ICE_CANDIDATE: 'ice.candidate',
  TRACK_MUTE: 'track.mute',
  /**
   * Declares what a track being published is *of*.
   *
   * Needed because WebRTC has no such concept, and a page can't choose the
   * `MediaStream` or `MediaStreamTrack` id the SDP will carry; both are
   * read-only. Without this the SFU can only guess the source from codec
   * kind, and that can't tell a screen share from a camera.
   */
  TRACK_PUBLISH: 'track.publish',
  SUBSCRIPTION_UPDATE: 'subscription.update',
  PING: 'ping',
} as const;

export const ServerMessageType = {
  ROOM_JOINED: 'room.joined',
  ROOM_LEFT: 'room.left',
  PARTICIPANT_JOINED: 'participant.joined',
  PARTICIPANT_LEFT: 'participant.left',
  TRACK_PUBLISHED: 'track.published',
  TRACK_UNPUBLISHED: 'track.unpublished',
  TRACK_MUTED: 'track.muted',
  TRACK_UNMUTED: 'track.unmuted',
  SDP_OFFER: 'sdp.offer',
  SDP_ANSWER: 'sdp.answer',
  ICE_CANDIDATE: 'ice.candidate',
  CONNECTION_STATE: 'connection.state',
  ERROR: 'error',
  PONG: 'pong',
} as const;

export type SignalingErrorCode =
  | 'INVALID_TOKEN'
  | 'TOKEN_EXPIRED'
  | 'TOKEN_REVOKED'
  | 'UNAUTHORIZED'
  | 'ROOM_NOT_FOUND'
  | 'ROOM_FULL'
  | 'INVALID_MESSAGE'
  | 'INVALID_MESSAGE_TYPE'
  | 'PARTICIPANT_NOT_FOUND'
  | 'NOT_IN_ROOM'
  | 'PERMISSION_DENIED'
  | 'ORIGIN_NOT_ALLOWED'
  | 'RATE_LIMITED'
  | 'USAGE_LIMIT_EXCEEDED'
  | 'NO_RTC_CAPACITY'
  | 'RTC_SERVER_UNREACHABLE'
  | 'NEGOTIATION_FAILED'
  | 'NEGOTIATION_GLARE';

export interface ServerTrack {
  trackId: string;
  kind: 'audio' | 'video';
  source: string;
  muted: boolean;
  simulcast: boolean;
  layers?: string[];
}

export interface ServerParticipant {
  id: string;
  tracks?: ServerTrack[];
}

export type ServerMessage =
  | { type: typeof ServerMessageType.ROOM_JOINED; roomId: string; participants: ServerParticipant[]; rtcServer?: string; region?: string }
  | { type: typeof ServerMessageType.ROOM_LEFT; roomId: string }
  | { type: typeof ServerMessageType.PARTICIPANT_JOINED; participant: ServerParticipant }
  | { type: typeof ServerMessageType.PARTICIPANT_LEFT; participant: ServerParticipant }
  | { type: typeof ServerMessageType.TRACK_PUBLISHED; participantId: string; track: ServerTrack }
  | { type: typeof ServerMessageType.TRACK_UNPUBLISHED; participantId: string; trackId: string }
  | { type: typeof ServerMessageType.TRACK_MUTED; participantId: string; trackId: string }
  | { type: typeof ServerMessageType.TRACK_UNMUTED; participantId: string; trackId: string }
  | { type: typeof ServerMessageType.SDP_OFFER; sdp: string }
  | { type: typeof ServerMessageType.SDP_ANSWER; sdp: string }
  | { type: typeof ServerMessageType.ICE_CANDIDATE; candidate: string; sdpMid?: string; sdpMLineIndex?: number; usernameFragment?: string }
  | { type: typeof ServerMessageType.CONNECTION_STATE; iceState: string; peerState: string }
  | { type: typeof ServerMessageType.ERROR; code: SignalingErrorCode; message: string }
  | { type: typeof ServerMessageType.PONG };

export type ClientMessage =
  | { type: typeof ClientMessageType.ROOM_JOIN; roomId?: string; region?: string }
  | { type: typeof ClientMessageType.ROOM_LEAVE }
  | { type: typeof ClientMessageType.SDP_ANSWER; sdp: string }
  | { type: typeof ClientMessageType.SDP_OFFER; sdp: string }
  | { type: typeof ClientMessageType.ICE_CANDIDATE; candidate: string; sdpMid?: string; sdpMLineIndex?: number; usernameFragment?: string }
  | { type: typeof ClientMessageType.TRACK_MUTE; trackId: string; muted: boolean }
  | { type: typeof ClientMessageType.TRACK_PUBLISH; trackId: string; source: 'camera' | 'microphone' | 'screenShare' }
  | { type: typeof ClientMessageType.SUBSCRIPTION_UPDATE; publisherId: string; trackId: string; layer: 'low' | 'medium' | 'high' | 'auto' }
  | { type: typeof ClientMessageType.PING };

/**
 * Error codes meaning "the credential is the problem", where reconnecting
 * on the same token can't possibly help.
 *
 * The split drives reconnect behaviour. Everything else is worth retrying
 * with backoff; these want a fresh token from the application's backend
 * first.
 */
export const FATAL_ERROR_CODES: ReadonlySet<SignalingErrorCode> = new Set([
  'INVALID_TOKEN',
  'TOKEN_EXPIRED',
  // Someone killed this credential on purpose. Retrying re-presents the
  // same dead token; the application's backend has to mint a new one.
  'TOKEN_REVOKED',
  'UNAUTHORIZED',
  'ROOM_NOT_FOUND',
  'PERMISSION_DENIED',
  // Reconnecting cannot help: the page's origin is not on the project's
  // allow-list, and that is changed in the dashboard, not by retrying.
  'ORIGIN_NOT_ALLOWED',
  // Not a credential problem, but just as terminal, and it belongs here
  // for the reconnect behaviour rather than the reason. The account is out
  // of included minutes: unlike RATE_LIMITED there is no window to wait
  // out, so backing off and retrying would spin against a wall until the
  // attempt budget ran out. A fresh token would not help either — the
  // limit is on the account, not the token.
  'USAGE_LIMIT_EXCEEDED',
]);

File: sdks/flutter/raven_rtc/lib/src/internal/protocol.dart
/// Raven's signaling wire protocol, as the Flutter client sees it.
///
/// Mirrors `packages/sdk/src/internal/signaling/protocol.ts` and
/// `apps/api/src/modules/signaling/signaling.constants.ts`. Written out
/// here rather than generated: three client implementations (web, React
/// Native, Flutter) speak this, and the contract has to be readable on its
/// own in each language.
///
/// If you change a message here, change it in the other two and in the
/// server. Nothing at build time will catch a drift — only the
/// integration tests will.
library;

import 'dart:convert';

/// Messages this client sends.
abstract final class ClientMessageType {
  static const roomJoin = 'room.join';
  static const roomLeave = 'room.leave';
  static const sdpAnswer = 'sdp.answer';
  static const sdpOffer = 'sdp.offer';
  static const iceCandidate = 'ice.candidate';
  static const trackMute = 'track.mute';

  /// Declares what a track being published is *of*.
  ///
  /// WebRTC carries no notion of source, and an application cannot choose
  /// the media-stream or track id that ends up in the SDP. Without this
  /// the server could only infer source from codec kind, which cannot
  /// tell a screen share from a camera.
  static const trackPublish = 'track.publish';
  static const subscriptionUpdate = 'subscription.update';
  static const ping = 'ping';
}

/// Messages the server sends.
abstract final class ServerMessageType {
  static const roomJoined = 'room.joined';
  static const roomLeft = 'room.left';
  static const participantJoined = 'participant.joined';
  static const participantLeft = 'participant.left';
  static const trackPublished = 'track.published';
  static const trackUnpublished = 'track.unpublished';
  static const trackMuted = 'track.muted';
  static const trackUnmuted = 'track.unmuted';
  static const sdpOffer = 'sdp.offer';
  static const sdpAnswer = 'sdp.answer';
  static const iceCandidate = 'ice.candidate';
  static const connectionState = 'connection.state';
  static const error = 'error';
  static const pong = 'pong';
}

/// Signaling error codes, as the server names them.
abstract final class SignalingErrorCode {
  static const invalidToken = 'INVALID_TOKEN';
  static const tokenExpired = 'TOKEN_EXPIRED';
  static const unauthorized = 'UNAUTHORIZED';
  static const roomNotFound = 'ROOM_NOT_FOUND';
  static const roomFull = 'ROOM_FULL';
  static const permissionDenied = 'PERMISSION_DENIED';
  static const rateLimited = 'RATE_LIMITED';
  static const noRtcCapacity = 'NO_RTC_CAPACITY';
  static const rtcServerUnreachable = 'RTC_SERVER_UNREACHABLE';
  static const negotiationFailed = 'NEGOTIATION_FAILED';
  static const negotiationGlare = 'NEGOTIATION_GLARE';

  /// Codes that mean the credential itself is the problem, so reconnecting
  /// with the same token cannot help.
  ///
  /// The distinction drives reconnect behaviour: everything else is worth
  /// retrying with backoff, these need a fresh token from the
  /// application's backend first.
  static const fatal = <String>{
    invalidToken,
    tokenExpired,
    unauthorized,
    roomNotFound,
    permissionDenied,
  };
}

/// A track the server says someone is publishing.
class ServerTrack {
  const ServerTrack({
    required this.trackId,
    required this.kind,
    required this.source,
    required this.muted,
    required this.simulcast,
    this.layers = const [],
  });

  factory ServerTrack.fromJson(Map<String, dynamic> json) => ServerTrack(
        trackId: json['trackId'] as String? ?? '',
        kind: json['kind'] as String? ?? 'video',
        source: json['source'] as String? ?? 'unknown',
        muted: json['muted'] as bool? ?? false,
        simulcast: json['simulcast'] as bool? ?? false,
        layers: (json['layers'] as List<dynamic>?)
                ?.map((layer) => layer as String)
                .toList(growable: false) ??
            const [],
      );

  final String trackId;

  /// `'audio'` or `'video'` — the codec kind, not the source.
  final String kind;

  /// `'camera'`, `'microphone'`, `'screenShare'`, or something this
  /// version does not know about.
  final String source;
  final bool muted;
  final bool simulcast;

  /// Layers actually being sent, when simulcast. Empty otherwise.
  final List<String> layers;
}

/// A participant the server says is in the room.
class ServerParticipant {
  const ServerParticipant({required this.id, this.tracks = const []});

  factory ServerParticipant.fromJson(Map<String, dynamic> json) =>
      ServerParticipant(
        id: json['id'] as String? ?? '',
        tracks: (json['tracks'] as List<dynamic>?)
                ?.map((track) =>
                    ServerTrack.fromJson(track as Map<String, dynamic>))
                .toList(growable: false) ??
            const [],
      );

  final String id;

  /// What this participant is already publishing.
  ///
  /// Present in `room.joined` so a client joining a call in progress can
  /// render the room in one pass, rather than showing an empty grid and
  /// filling it in from a stream of events it has to distinguish from
  /// genuinely new ones.
  final List<ServerTrack> tracks;
}

/// What the server reports at the moment of joining.
class JoinedPayload {
  const JoinedPayload({
    required this.roomId,
    required this.participants,
    this.rtcServer,
    this.region,
  });

  factory JoinedPayload.fromJson(Map<String, dynamic> json) => JoinedPayload(
        roomId: json['roomId'] as String? ?? '',
        participants: (json['participants'] as List<dynamic>?)
                ?.map((participant) => ServerParticipant.fromJson(
                    participant as Map<String, dynamic>))
                .toList(growable: false) ??
            const [],
        rtcServer: json['rtcServer'] as String?,
        region: json['region'] as String?,
      );

  final String roomId;
  final List<ServerParticipant> participants;

  /// The RTC server's *name*, for diagnostics. Never its address — a
  /// client that learned an SFU's address could connect to it directly,
  /// and then the media plane could not be changed without breaking that
  /// client.
  final String? rtcServer;
  final String? region;
}

/// An ICE candidate, in the shape both directions of the protocol use.
class IceCandidatePayload {
  const IceCandidatePayload({
    required this.candidate,
    this.sdpMid,
    this.sdpMLineIndex,
    this.usernameFragment,
  });

  factory IceCandidatePayload.fromJson(Map<String, dynamic> json) =>
      IceCandidatePayload(
        candidate: json['candidate'] as String? ?? '',
        sdpMid: json['sdpMid'] as String?,
        sdpMLineIndex: json['sdpMLineIndex'] as int?,
        usernameFragment: json['usernameFragment'] as String?,
      );

  final String candidate;
  final String? sdpMid;
  final int? sdpMLineIndex;
  final String? usernameFragment;

  Map<String, dynamic> toJson() => {
        'type': ClientMessageType.iceCandidate,
        'candidate': candidate,
        if (sdpMid != null) 'sdpMid': sdpMid,
        if (sdpMLineIndex != null) 'sdpMLineIndex': sdpMLineIndex,
        if (usernameFragment != null) 'usernameFragment': usernameFragment,
      };
}

/// Reads the claims out of a Raven RTC token.
///
/// The payload is readable, not secret — the same information the server
/// will act on. It is decoded, never trusted: the server re-verifies the
/// signature, so anything a client changed here only changes which room it
/// *asks* for.
Map<String, dynamic>? decodeTokenClaims(String token) {
  final parts = token.split('.');
  if (parts.length != 3) return null;

  try {
    // base64Url tolerates the unpadded form the JWT spec mandates, so no
    // padding fix-up is needed — unlike plain `base64`, which throws.
    final decoded =
        utf8.decode(base64Url.decode(base64Url.normalize(parts[1])));
    final parsed = jsonDecode(decoded);
    return parsed is Map<String, dynamic> ? parsed : null;
  } catch (_) {
    return null;
  }
}
